File: binder/C++_test/src/test_client.cpp
/*consult frameworks/av/media/medaserver/Main_mediaserver.cpp */

#define LOG_TAG "HelloService"
//#define LOG_NDEBUG 0

#include <fcntl.h>
#include <sys/prctl.h>
#include <sys/wait.h>
#include <binder/IPCThreadState.h>
#include <binder/ProcessState.h>
#include <binder/IServiceManager.h>
#include <cutils/properties.h>
#include <utils/Log.h>

using namespace android;


/* ./test_client sayhello
 *  ./test_client hello <name>
 *
 * */
int main(int argc, char **argv)
{
    int cnt = 0;

    if (argc < 2) {
        ALOGI("usage:\n");
        ALOGI("%s hello or goodbye\n", argv[0]);
        ALOGI("%s hello or goodbye <name>\n", argv[0]);
        return -1;
    }
    
    /*getService*/
    /*open driver, mmap*/
    sp<ProcessState> proc(ProcessState::self());

    /*获得BpServiceManager*/
    sp<IServiceManager> sm = defaultServiceManager();

    sp<IBinder> binder = sm->getService(String16("hello"));
    if (binder == 0) {
        ALOGI("cannot get hello service");
        return -1;
    }
    sp<IHelloService> service = interface_cast<IHelloService>(binder);
    
    /*调用Service的函数*/
    if (argc < 2) {
        service->sayhello();
        ALOGI("client call sayhello");
    } else { 
        cnt = service->sayhello_to(argv[2]);
        ALOGI("client call sayhello_to, cnt = %d", cnt);
    }

    return 0;
}

File: 0.Binder_System/Java_test/src/TestClient.java
/*
 * 1.getService
 * 2. call Service's sayhello sayhello_to
*/

/*
 * test_client <hello|goodbye> name
 * */
import android.util.Slog;
import android.os.ServiceManager;
import android.os.IBinder;

public class TestClient {
    private static final String TAG = "HelloService";
    
    public static void main(String args[]) {
        if (args.length == 0) {
            System.out.println("Usage:need params:<hello|goodbye> name");
            return;
        }

        if (args[0].equals("hello")) {
            IBinder binder = ServiceManager.getService("hello");
            if (binder == null) {
                System.out.println("cannot get hello Service");
                Slog.i(TAG, "cannot get hello Service");
                return;
            }
                
            IHelloService svr = IHelloService.Stub.asInterface(binder);

            if(args.length == 1) {
                try {
                    svr.sayhello();
                    System.out.println("call sayhello");
                    Slog.i(TAG, "call sayhello");
                } catch (Exception e) {

                }
            } else {
                try {
                    int cnt = svr.sayhello_to(args[1]);
                    System.out.println("call sayhello_to "+args[1]+":cnt2 = "+cnt);
                    Slog.i(TAG, "call sayhello_to "+args[1]+":cnt2 = "+cnt);
                } catch (Exception e) {

                }
            }
        }
    }
}
